src/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp

```diff
@@ -191,6 +191,12 @@ static size_t WriteData(char* ptr, size_t size, size_t nmemb, void* userdata)
         }
 
         HttpResponse* response = context->m_response;
+        auto& headersHandler = context->m_request->GetHeadersReceivedEventHandler();
```

Comment:
to be honest, this is the wackiest part of this PR (the rest is quite nice).

However,
libcurl does not provide a way to receive HTTP2 headers in a raw hpack H2 event format.
libcurl always transforms HTTP2 headers into http 1.1 like headers.

Also libcurl does not have a callback to notify when all headers are received or when a single payload-less H2 frame handling is done. (technically we receive H2 frame with "connection" headers, but libcurl presents it to us like a regular http1.1 response headers).

So here, 
it is not "on headers received" but "headers maybe received but now you are receiving a payload" which is quite ugly/wacky, but it is the best I managed to get from libCurl.

The alternative would be: using a different HTTP client or [CURLOPT_CONNECT_ONLY](https://curl.se/libcurl/c/CURLOPT_CONNECT_ONLY.html) (essentially asking libcurl to not perform http2->http1.1 headers transform and implementing a significant part of http protocol ourselves).